datafusion-physical-expr/src/expressions/approx_percentile_cont_with_weight.rs

```diff
@@ -0,0 +1,154 @@
+        debug_assert_eq!(expr.len(), 3);
+
+        let sub_expr = vec![expr[0].clone(), expr[2].clone()];
+        let approx_percentile_cont =
+            ApproxPercentileCont::new(sub_expr, name, return_type)?;
+
+        Ok(Self {
+            approx_percentile_cont,
+            column_expr: expr[0].clone(),
+            weight_expr: expr[1].clone(),
+        })
+    }
+}
+
+impl AggregateExpr for ApproxPercentileContWithWeight {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn field(&self) -> Result<Field> {
+        self.approx_percentile_cont.field()
+    }
+
+    #[allow(rustdoc::private_intra_doc_links)]
+    /// See [`TDigest::to_scalar_state()`] for a description of the serialised
+    /// state.
+    fn state_fields(&self) -> Result<Vec<Field>> {
+        self.approx_percentile_cont.state_fields()
+    }
+
+    fn expressions(&self) -> Vec<Arc<dyn PhysicalExpr>> {
+        vec![self.column_expr.clone(), self.weight_expr.clone()]
+    }
+
+    fn create_accumulator(&self) -> Result<Box<dyn Accumulator>> {
+        let approx_percentile_cont_accumulator =
+            self.approx_percentile_cont.create_plain_accumulator()?;
+        let accumulator = ApproxPercentileWithWeightAccumulator::new(
+            approx_percentile_cont_accumulator,
+        );
+        Ok(Box::new(accumulator))
+    }
+
+    fn name(&self) -> &str {
+        self.approx_percentile_cont.name()
+    }
+}
+
+#[derive(Debug)]
+pub struct ApproxPercentileWithWeightAccumulator {
+    approx_percentile_cont_accumulator: ApproxPercentileAccumulator,
+}
+
+impl ApproxPercentileWithWeightAccumulator {
+    pub fn new(approx_percentile_cont_accumulator: ApproxPercentileAccumulator) -> Self {
+        Self {
+            approx_percentile_cont_accumulator,
+        }
+    }
+}
+
+impl Accumulator for ApproxPercentileWithWeightAccumulator {
+    fn state(&self) -> Result<Vec<ScalarValue>> {
+        self.approx_percentile_cont_accumulator.state()
+    }
+
+    fn update_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
+        debug_assert_eq!(
+            values.len(),
+            2,
+            "invalid number of values in batch percentile update"
+        );
+        let means = &values[0];
+        let weights = &values[1];
+        debug_assert_eq!(
+            means.len(),
+            weights.len(),
+            "invalid number of values in means and weights"
+        );
+        let means_f64 = ApproxPercentileAccumulator::convert_to_ordered_float(means)?;
+        let weights_f64 = ApproxPercentileAccumulator::convert_to_ordered_float(weights)?;
+        let mut digests: Vec<TDigest> = vec![];
+        for (mean, weight) in means_f64.iter().zip(weights_f64.iter()) {
+            digests.push(TDigest::new_with_centroid(
+                DEFAULT_MAX_SIZE,
+                Centroid::new(*mean, *weight),
+            ))
+        }
+        self.approx_percentile_cont_accumulator
+            .merge_digests(&digests);
```

Comment:
IMO, still use `approx_percentile_cont_accumulator.merge_digests`, why not change the struct `[approx_percentile_cont.rs]` to support `weight `. If i miss something plz tell me.

Reply:
@Ted-Jiang good question.

I separate them out because their input arguments are slightly different, one is `(column_expr, percentile)`, another is `(column_expr, weight_expr, percentile)`. 

ps: NOT using `(column_expr, percentile, weight_expr)` because I am following Presto/Trino

https://github.com/apache/arrow-datafusion/blob/cda79e1b8a9e27acc604379ad4de18dc00d11275/datafusion/src/physical_plan/aggregates.rs#L359-L376

---

that says, after merge, we could simplify `approx_percentile_cont` by moving all its logic to `approx_percentile_cont_with_weight` and add an optimization rule to rewrite `approx_percentile_cont(x, p)` to `approx_percentile_cont_with_weight(x, 1, p)`